### PR TITLE
Support Lysandros' new `tstrings` branch of `cpython`

### PR DESCRIPTION
**Pull request by davepeck**
This branch uses @lysnikolaou's in-progress [`tstrings` branch of cpython](https://github.com/lysnikolaou/cpython/tree/tstrings). 

Thanks to @koxudaxi for getting docker images for @lysnikolaou's branch together!

Because the `tstrings` branch is very much in progress, we `@pytest.mark.skipif(...)` quite a number of tests. We've documented the following divergences so far:

- [ ] Implicit concatenation with t-strings is not supported (`MISSING_IMPLICIT_CONCAT` -- search for commented-out tests when fixing)
- [ ] Debug specifier behavior requires a `!s` conv when a format spec is also provided (that is, `t"{expr=}"` should become `t"expr={expr!r}"` but `t"{expr=:fmt}"` should become `t"expr={expr!s:fmt}"`). (`BUG_DEBUG_SPECIFIER_WITH_FMT`)
- [x] Format specifications do not support nested interpolations (e.g. `t"Value: {value:.{precision}f}"` should evaluate format_spec to ".2f"). (`BUG_NESTED_FORMAT_SPEC`)
- [x] Constructing `Interpolation`s with invalid values results in a segfault (`BUG_INTERPOLATION_CONSTRUCTOR_SEGFAULT`)
- [x] Template strings raise MemoryError when containing more than 16 expressions (`BUG_MANY_EXPRESSIONS` )
- [x]  Constructing Interpolations ignore conv argument (BUG_INTERPOLATION_CONSTRUCTOR_CONV)
- [x] `Template` and `Interpolation` do not currently support `__hash__()`. (`MISSING_TEMPLATE_HASH`)
- [x] `Template.args` is not currently interleaved. (`MISSING_INTERLEAVING`)
- [x] `Template.__init__()` has the wrong signature, should be `*args: str | Interpolation`. (`BUG_TEMPLATE_CONSTRUCTOR` -- search for commented-out tests when fixing)
- [x] Constant templates like `t"Hello"` are turned into `Constant`, not `Template`. (`BUG_CONSTANT_TEMPLATE`) 🎉
- [x] A t-string with a single interpolation like `t"{42}"` evaluates to an `Interpolation` directly, not a `Template` with an `Interpolation` in its `args`. (`BUG_SINGLE_INTERPOLATION`) 🎉
- [x] `Template` does not currently support `__add__()` or `__radd()__`. (`MISSING_TEMPLATE_ADD_RADD`)
- [x] `Template` does not currently support `__eq__()` as described in the PEP. (`MISSING_TEMPLATE_EQ`)
- [x] `Interpolation` does not yet support positional pattern matching via `__match_args__ = ('value', 'expr', 'conv', 'format_spec')`. (`BUG_INTERPOLATION_MATCH_ARGS`)
- [x] t-string  does not currently support debug specifier `t"{name=}"` (`_BUG_DEBUG_SPECIFIER`)
- [x] t-string  raise `f-string` syntax error messages (`_INCORRECT_SYNTAX_ERROR_MESSAGE`)
As the `tstrings` implementation gets closer, we'll update this PR and un-skip the relevant bugs. ⏳

### CODE REVIEW

File: pep/test.py [hunk 1]
```diff
@@ -159,73 +186,70 @@ def test_add_str_template_2():
...
+    _INCORRECT_SYNTAX_ERROR_MESSAGE,
+    reason="Template syntax error message not implemented",
+)
+def test_syntax_error_1():
+    with pytest.raises(
+        SyntaxError, match="t-string: valid expression required before '}'"
+    ):
+        # Use exec to avoid syntax error in the test itself
+        exec('t"hello {}"')
+        # Now, we got "f-string: valid expression required before '}'"
```
Comment (koxudaxi) [inline pep/test.py:255]: I think this is not a blocker for merge. But I found it and added it.

Comment (davepeck) [reply pep/test.py:255]: Oh interesting, that's a good one.

**Comment by lysnikolaou** [comment]
> * [x]  t-string  does not currently support debug specifier `t"{name=}"` (`_BUG_DEBUG_SPECIFIER`)

Fixed as well!

**Comment by koxudaxi** [comment]
@lysnikolaou 
I found the bug of `MISSING_TEMPLATE_ADD_RADD `.

```python
 ./python.exe
Python 3.14.0a1+ (heads/tstrings:94371bd985, Oct 23 2024, 17:04:52) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from templatelib import Template
>>> name = "world"
... template = "hello " + t"there, {name}!"
... assert isinstance(template, Template)
...
... assert len(template.args) == 4
... assert template.args[0] == "hello "
... assert template.args[1] == "there, "
...
... assert len(template.args) == 3
...
Traceback (most recent call last):
  File "<python-input-1>", line 9, in <module>
    assert len(template.args) == 3
           ^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

**Comment by koxudaxi** [comment]
`_BUG_DEBUG_SPECIFIER` and  `MISSING_TEMPLATE_EQ ` are looks good.
I will update the flag after docker image will be pushed.

**Comment by lysnikolaou** [comment]
> @lysnikolaou
> I found the bug of `MISSING_TEMPLATE_ADD_RADD `

Yes, right now, we only have a very naive implementation for __add__. This'll be resolved once interleaving is implemented as well.



> [x] `Interpolation` does not yet support positional pattern matching via `__match_args__ = ('value', 'expr', 'conv', 'format_spec')`. (`BUG_INTERPOLATION_MATCH_ARGS`)

Done!

**Comment by lysnikolaou** [comment]
> * t-string  raise `f-string` syntax error messages (`_INCORRECT_SYNTAX_ERROR_MESSAGE`)
>   As the `tstrings` implementation gets closer, we'll update this PR and un-skip the relevant bugs. ⏳

Fixed, too!

**Comment by davepeck** [comment]
PS There's a python implementation of `interleave(*args: str | Interpolation) -> tuple[str | Interpolation]` here:

https://github.com/davepeck/pep750-examples/blob/main/pep/__init__.py#L86

**Comment by davepeck** [comment]
First of all: 🎉

Second of all: added two more to the list above: `_MISSING_TEMPLATE_HASH` and `_BUG_TEMPLATE_CONSTRUCTOR`

**Comment by koxudaxi** [comment]
I added `BUG_NESTED_FORMAT_SPEC` and the test

**Comment by davepeck** [comment]
Added `_BUG_INTERPOLATION_CONSTRUCTOR_SEGFAULT`. Some attempts to construct `Interpolation`s with invalid arguments results in a segv at the moment.

**Comment by davepeck** [comment]
Added a number of explicit tests of both the `Interpolation` and `Template` constructors.

Testing the `Template` constructor also tests automated interleaving; I've added a number of tests there.

The most interesting interleaving test is this one:

```python
def test_template_constructor_interleaving_empty():
    """Test probably the most interesting corner case with interleaving."""
    # The PEP 750 specification requires that `Template.args` contains
    # one more string than interpolations. If there are no interpolations,
    # we must therefore have one empty string in `Template.args`.
    template = Template()
    expected = ("",)
    assert tuple(template.args) == expected
```

This behavior is implied by the PEP, but it's a corner case we should probably explicitly call out in the "interleaving" section.

**Comment by koxudaxi** [comment]
I added `_BUG_MANY_EXPRESSIONS`.

**Comment by lysnikolaou** [comment]
`BUG_NESTED_FORMAT_SPEC`, `BUG_INTERPOLATION_CONSTRUCTOR_SEGFAULT` and `BUG_MANY_EXPRESSIONS` should be fixed now.

**Comment by koxudaxi** [comment]
@lysnikolaou 
I tested the latest commit.
The Interpolation object doesn't have a `conv` value.
```python
>>> from templatelib import *
>>> i1 = Interpolation(42, "i1", "a", ",.2f")
>>> i1.conv == "a"
False
>>> i1.conv
>>> 
```
https://github.com/davepeck/pep750-examples/actions/runs/11519564794/job/32068860023#step:6:127

**Comment by koxudaxi** [comment]
@lysnikolaou 
I added  `BUG_INTERPOLATION_CONSTRUCTOR_IGNORE_CONV` for the issue.

**Comment by davepeck** [comment]
@lysnikolaou Added `_BUG_DEBUG_SPECIFIER_WITH_FMT` (description above). We're getting into the weeds at this point. :-)
